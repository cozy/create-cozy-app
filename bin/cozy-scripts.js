#! /usr/bin/env node

'use strict'

const commander = require('commander')
const pkg = require('../package.json')
const colorize = require('../utils/_colorize.js')
const CTS = require('../utils/constants.js')
const regexpReplacer = require('../utils/regexpReplacer')
const getWebpackConfigs = require('../scripts/config')

let actionName
const program = new commander.Command(pkg.name)
  .description(
    `
    Available actions:
    - build: build your application once (Webpack)
    - watch: build your application and listen to changes to rebuild it automatically (Webpack)
    - start: build your application and serve it inside a Cozy using Docker (with HMR)
    - test: run your tests using Jest
    - publish: run cozy-app-publish package (cf cozy-app-publish documentation)
    - release: run cozy-release to manage an app version release (cf cozy-release documentation)
    `
  )
  .version(pkg.version)
  .arguments('<action-name>')
  .usage(`${colorize.blue('<action-name>')} [options]`)
  .action(name => {
    actionName = name
  })
  .option('--browser', 'specify browser build target')
  .option('--debug', 'print more outputs for debugging')
  .option('--development', 'specify development build mode')
  .option('--fix', 'format automatically the code with eslint')
  .option('--no-hot', 'disables hot module reload (only for development)')
  .option(
    '--stack',
    'use cozy-stack with docker launch when using `cozy-scripts start`',
    false
  )
  .option(
    '--port  <portNumber>',
    'change the webpack dev server port (default 8888) (only for development)'
  )
  .option(
    '--host  <hostName>',
    'change the webpack dev server host name (default localhost) (only for development)'
  )
  .option('--mobile', 'specify mobile build target')
  .option('--production', 'specify production build mode')
  .option('--show-config', 'just print app final webpack config')
  .option(
    '--config <pathToFile>',
    'provide a custom cozy-scripts build config file path (relative to the application root directory). Use it only if you need custom build configuration using app.config.js with a custom location.'
  )
  .option(
    '--src-dir <pathToDirectory>',
    'provide the application source (`src`) directory path (relative to the application root directory)'
  )
  .option(
    '--build-dir <pathToDirectory>',
    'provide the application `build` directory path (relative to the application root directory) to build the application into'
  )
  .option(
    '--manifest <pathToFile>',
    'provide the application manifest file path (relative to the application root directory)'
  )
  .option(
    '--analyzer',
    'open an analyzer with an interactive treemap visualization of the contents of all builds'
  )
  .option(
    '--devtool',
    'Configure the devtool used. Use false to deactivate completely'
  )
  .parse(process.argv)

// build mode and target computing (overwritten by NODE_ENV)
const options = {
  mode:
    (program.production && 'production') ||
    (program.development && 'development') ||
    (actionName === 'build' && 'production') ||
    'development',
  target:
    (program.browser && 'browser') || (program.mobile && 'mobile') || 'browser',
  // all arguments passed to the command (we remove the main command name)
  cliArgs: process.argv.slice(3)
}

const getEnvVarsFromCLIArgs = program => {
  const env = {}
  // program property, environment variable name, content to set
  ;[
    ['hot', CTS.HOT, true],
    ['port', CTS.PORT, program.port],
    ['host', CTS.HOST, program.host],
    ['fix', CTS.ESLINT_FIX, true],
    ['debug', CTS.DEBUG, true],
    ['analyzer', CTS.ANALYZER, true],
    ['config', CTS.CONFIG, program.config],
    ['srcDir', CTS.SRC_DIR, program.srcDir],
    ['buildDir', CTS.BUILD_DIR, program.buildDir],
    ['manifest', CTS.MANIFEST, program.manifest],
    ['devtool', CTS.DEVTOOL, program.devtool]
  ].map(toDefine => {
    if (program[toDefine[0]] !== undefined) {
      env[toDefine[1]] = toDefine[2]
    }
  })
  return env
}

const envVars = getEnvVarsFromCLIArgs(program)
Object.assign(process.env, envVars)

if (program.showConfig) {
  console.log(JSON.stringify(getWebpackConfigs(options), regexpReplacer, 2))
} else {
  const availableScripts = [
    'build',
    'watch',
    'start',
    'test',
    'publish',
    'release',
    'lint',
    'check-locales',
    'stackDocker',
    'deploy'
  ]

  if (availableScripts.includes(actionName)) {
    if (actionName === 'start') {
      // specific to this action
      options.stack = program.stack // specific behaviour of --no-* options
    }
    const scriptPath = `../scripts/${actionName}`
    const script = require(scriptPath)
    script(options)
  } else {
    if (!actionName) {
      console.error(
        'Use `--help` option to get more informations about cozy-scripts usage.'
      )
    } else {
      console.error(
        `cozy-scripts: unknown command ${colorize.cyan(actionName)}`
      )
    }
  }
}
