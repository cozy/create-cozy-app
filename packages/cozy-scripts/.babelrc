 {
  "presets": ["cozy-app", "@babel/typescript", "@babel/env"]
}
