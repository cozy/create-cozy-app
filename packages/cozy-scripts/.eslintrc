{
  "extends": ["cozy-app"],
  "root": true
}
