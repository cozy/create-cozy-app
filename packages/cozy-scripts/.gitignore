# NPM
node_modules/
npm-debug.log
yarn-error.log


# Build
build/

# Coverage
coverage/


# Locales
src/locales/*
!src/locales/en.json


# transifex
+.transifexrc


# Coverage
coverage/


# OS generated files
.DS_Store
.DS_Store?
._*
.Spotlight-V100
.Trashes
ehthumbs.db
Thumbs.db
desktop.ini


# Editors / IDEs
.floo
.flooignore


# Mobile
mobile/hooks
mobile/platforms
mobile/plugins
mobile/www
mobile/keys


# fastlane
mobile/fastlane/README.md
mobile/fastlane/report.xml
mobile/fastlane/Error*.png
mobile/fastlane/Preview.html
mobile/Preview.html


# Default
# /!\ KEEP THIS SECTION THE LAST ONE
!.gitkeep
