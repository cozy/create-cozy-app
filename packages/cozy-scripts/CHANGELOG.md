# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [5.8.0](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@5.7.0...cozy-scripts@5.8.0) (2021-07-08)


### Features

* Support cozy.localhost ([6fccd13](https://github.com/cozy/create-cozy-app/commit/6fccd13))





# [5.7.0](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@5.6.0...cozy-scripts@5.7.0) (2021-05-10)


### Features

* Update jest and babel-jest ([62b41ac](https://github.com/cozy/create-cozy-app/commit/62b41ac))





# [5.6.0](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@5.5.0...cozy-scripts@5.6.0) (2021-03-30)


### Bug Fixes

* Be more precise on styles mock ([4b77c2c](https://github.com/cozy/create-cozy-app/commit/4b77c2c))


### Features

* Remove transifex template ([d058c6f](https://github.com/cozy/create-cozy-app/commit/d058c6f))





# [5.5.0](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@5.4.4...cozy-scripts@5.5.0) (2021-03-30)


### Features

* Remove tx installation ([2ae2eb8](https://github.com/cozy/create-cozy-app/commit/2ae2eb8))
* Upgrade node, remove fastfinish and travisorg ([6f2eff1](https://github.com/cozy/create-cozy-app/commit/6f2eff1))





## [5.4.4](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@5.4.3...cozy-scripts@5.4.4) (2021-03-23)


### Bug Fixes

* Add settings permission  ([fe962f4](https://github.com/cozy/create-cozy-app/commit/fe962f4))





## [5.4.3](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@5.4.2...cozy-scripts@5.4.3) (2021-03-22)


### Bug Fixes

* Import cozy-ui CSS as early as possible ([6e32796](https://github.com/cozy/create-cozy-app/commit/6e32796)), closes [/github.com/cozy/cozy-ui/issues/1774#issuecomment-803967934](https://github.com//github.com/cozy/cozy-ui/issues/1774/issues/issuecomment-803967934)





## [5.4.2](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@5.4.1...cozy-scripts@5.4.2) (2020-12-22)

**Note:** Version bump only for package cozy-scripts





## [5.4.1](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@5.4.0...cozy-scripts@5.4.1) (2020-11-13)


### Bug Fixes

* ReactHotLoader is used in webpack config react ([87eccab](https://github.com/cozy/create-cozy-app/commit/87eccab))





# [5.4.0](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@5.3.1...cozy-scripts@5.4.0) (2020-11-04)


### Features

* Add git-directory-deploy since it is necessary for cs deploy ([e8994dc](https://github.com/cozy/create-cozy-app/commit/e8994dc))
* Error were not correctly logged ([852055d](https://github.com/cozy/create-cozy-app/commit/852055d))
* Execute cozy-app-publish from local node_modules via npx ([cf38d80](https://github.com/cozy/create-cozy-app/commit/cf38d80)), closes [#1362](https://github.com/cozy/create-cozy-app/issues/1362)
* Execute cozy-release from local node_modules via npx ([bf1d560](https://github.com/cozy/create-cozy-app/commit/bf1d560))





## [5.3.1](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@5.3.0...cozy-scripts@5.3.1) (2020-10-23)

**Note:** Version bump only for package cozy-scripts





# [5.3.0](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@5.2.0...cozy-scripts@5.3.0) (2020-10-15)


### Bug Fixes

* Don't be too strict in the version ([68af134](https://github.com/cozy/create-cozy-app/commit/68af134))


### Features

* Upgrade cozy's lib ([7f6e744](https://github.com/cozy/create-cozy-app/commit/7f6e744))
* Upgrade react ([52dbca7](https://github.com/cozy/create-cozy-app/commit/52dbca7))





# [5.2.0](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@5.1.3...cozy-scripts@5.2.0) (2020-10-15)


### Features

* Add deploy command using git-directory-deploy ([d153f3c](https://github.com/cozy/create-cozy-app/commit/d153f3c))





## [5.1.3](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@5.1.2...cozy-scripts@5.1.3) (2020-10-06)

**Note:** Version bump only for package cozy-scripts





## [5.1.2](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@5.1.1...cozy-scripts@5.1.2) (2020-09-29)


### Bug Fixes

* Font from CSS could not be loaded ([84242d9](https://github.com/cozy/create-cozy-app/commit/84242d9))





## [5.1.1](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@5.1.0...cozy-scripts@5.1.1) (2020-09-28)


### Bug Fixes

* Add relative node_modules in resolve.modules ([cb72c06](https://github.com/cozy/create-cozy-app/commit/cb72c06))





# [5.1.0](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@5.0.2...cozy-scripts@5.1.0) (2020-09-23)


### Features

* Cozy-scripts installs the stylint dependency ([83d96d5](https://github.com/cozy/create-cozy-app/commit/83d96d5))
* Share .stylintrc ([067a883](https://github.com/cozy/create-cozy-app/commit/067a883))





## [5.0.2](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@5.0.1...cozy-scripts@5.0.2) (2020-09-23)


### Bug Fixes

* Use cozy-scripts v5 ([cf6be83](https://github.com/cozy/create-cozy-app/commit/cf6be83))





## [5.0.1](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@5.0.0...cozy-scripts@5.0.1) (2020-09-15)


### Bug Fixes

* Prevent flaky builds ([fc95ebc](https://github.com/cozy/create-cozy-app/commit/fc95ebc))





# [5.0.0](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@4.4.0...cozy-scripts@5.0.0) (2020-09-08)


### Features

* Removes the stack in docker option ([e89ad78](https://github.com/cozy/create-cozy-app/commit/e89ad78))


### BREAKING CHANGES

* The --stack flag is no longer supported. As an
alternative, it's possible to directly run the stack either from source
or with Docker. The command to run the docker image is in the README.





# [4.4.0](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@4.3.3...cozy-scripts@4.4.0) (2020-08-31)


### Features

* Add check-locales script ([88e9d9e](https://github.com/cozy/create-cozy-app/commit/88e9d9e))





## [4.3.3](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@4.3.2...cozy-scripts@4.3.3) (2020-08-31)

**Note:** Version bump only for package cozy-scripts





## [4.3.2](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@4.3.1...cozy-scripts@4.3.2) (2020-08-25)


### Bug Fixes

* Deactivate hot reload for production ([d65fb1c](https://github.com/cozy/create-cozy-app/commit/d65fb1c))





## [4.3.1](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@4.3.0...cozy-scripts@4.3.1) (2020-08-20)


### Bug Fixes

* Add allowed hosts to fix "Invalid Host Header" error ([69093d6](https://github.com/cozy/create-cozy-app/commit/69093d6))





# [4.3.0](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@4.2.1...cozy-scripts@4.3.0) (2020-07-28)


### Features

* Add babel-preset-cozy-app ([c1b0f64](https://github.com/cozy/create-cozy-app/commit/c1b0f64))
* Add BreakpointsProvider ([14ce917](https://github.com/cozy/create-cozy-app/commit/14ce917))
* Have babel-jest and jest at the same version ([3e021fa](https://github.com/cozy/create-cozy-app/commit/3e021fa))
* Import stylesheet and utilities ([805ed45](https://github.com/cozy/create-cozy-app/commit/805ed45))
* Update deps inside app template package.json ([b616b63](https://github.com/cozy/create-cozy-app/commit/b616b63))
* Use the newer [@babel](https://github.com/babel)/polyfill ([659644f](https://github.com/cozy/create-cozy-app/commit/659644f))





## [4.2.1](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@4.2.0...cozy-scripts@4.2.1) (2020-07-21)


### Bug Fixes

* Provide cozyclientjs in dev mode ([#1334](https://github.com/cozy/create-cozy-app/issues/1334)) ([9d6b6e5](https://github.com/cozy/create-cozy-app/commit/9d6b6e5))





# [4.2.0](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@4.1.0...cozy-scripts@4.2.0) (2020-07-21)


### Features

* Automatically provide the bar via ProvidePlugin if used through cozy.bar ([f30bef4](https://github.com/cozy/create-cozy-app/commit/f30bef4))





# [4.1.0](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@4.0.0...cozy-scripts@4.1.0) (2020-06-08)


### Bug Fixes

* Having sourcemap to true here broke cozy-ui ([3e5235a](https://github.com/cozy/create-cozy-app/commit/3e5235a))


### Features

* Add devtool flag to CLI ([1e2e41d](https://github.com/cozy/create-cozy-app/commit/1e2e41d))
* Add eslint-plugin-react-hooks as dependency ([010dcf1](https://github.com/cozy/create-cozy-app/commit/010dcf1))
* Use HashModuleIds only in production ([157e723](https://github.com/cozy/create-cozy-app/commit/157e723))





# [4.0.0](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@3.0.0...cozy-scripts@4.0.0) (2020-06-04)


### Bug Fixes

* No browser option passed to autoprefixer ([8c8ac64](https://github.com/cozy/create-cozy-app/commit/8c8ac64))


### BREAKING CHANGES

* Apps must declare a .browserslistrc

Put the content below in `.browserslistrc`:

```
["extends browserslist-config-cozy"]
```





# [3.0.0](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@1.15.0...cozy-scripts@3.0.0) (2020-04-07)


### Features

* Update dependencies ([1dbf5fa](https://github.com/cozy/create-cozy-app/commit/1dbf5fa))


### BREAKING CHANGES

* Jest is updated from v23 to v24





# [2.0.0](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@1.15.0...cozy-scripts@2.0.0) (2020-04-07)


### Features

* Update dependencies ([1dbf5fa](https://github.com/cozy/create-cozy-app/commit/1dbf5fa))


### BREAKING CHANGES

* Jest is updated from v23 to v24





# [1.15.0](https://github.com/cozy/create-cozy-app/compare/cozy-scripts@1.14.1...cozy-scripts@1.15.0) (2020-04-01)


### Features

* Move dependencies to cozy-scripts ([9b2aee5](https://github.com/cozy/create-cozy-app/commit/9b2aee5))





## [1.14.1](https://github.com/CPatchane/create-cozy-app/compare/cozy-scripts@1.14.0...cozy-scripts@1.14.1) (2020-03-31)


### Bug Fixes

* Include plugins folder in what is shipped to npm ([a9212e5](https://github.com/CPatchane/create-cozy-app/commit/a9212e5))





# [1.14.0](https://github.com/CPatchane/create-cozy-app/compare/cozy-scripts@1.13.2...cozy-scripts@1.14.0) (2020-03-31)


### Bug Fixes

* Consider full parts of path when including pictures ([3800cf3](https://github.com/CPatchane/create-cozy-app/commit/3800cf3))
* Do not override manifest content from locales ([1a1f1db](https://github.com/CPatchane/create-cozy-app/commit/1a1f1db))
* Remove vue ref ([272860b](https://github.com/CPatchane/create-cozy-app/commit/272860b))


### Features

* A stack is not started by default ([bc79e9d](https://github.com/CPatchane/create-cozy-app/commit/bc79e9d))
* Add config snap ([6f697d8](https://github.com/CPatchane/create-cozy-app/commit/6f697d8))
* Add VersionPlugin ([cc895a6](https://github.com/CPatchane/create-cozy-app/commit/cc895a6))
* Hot reload defaults to true ([11e8b5b](https://github.com/CPatchane/create-cozy-app/commit/11e8b5b))
* Remove deprecated standalone in favor of start ([491ddf7](https://github.com/CPatchane/create-cozy-app/commit/491ddf7))
* Remove preact specific configs ([7fb443e](https://github.com/CPatchane/create-cozy-app/commit/7fb443e))
* Remove vue option ([5a25d25](https://github.com/CPatchane/create-cozy-app/commit/5a25d25))


### Performance Improvements

* Speed up webpack dev ([e7f4b14](https://github.com/CPatchane/create-cozy-app/commit/e7f4b14))
* Use cheaper source maps for development ([e4e91bc](https://github.com/CPatchane/create-cozy-app/commit/e4e91bc))
