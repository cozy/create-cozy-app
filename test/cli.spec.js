/* eslint-env jest */

const path = require('path')
// to remove chalk colors escape codes and handling all testing environments
const stripAnsi = require('strip-ansi')
const CTS = require('../utils/constants.js')

const actualArgv = process.argv.slice(0) // a clone
const actualLog = console.log
const actualError = console.error
let currentScriptFile = null

// mocking
;['build', 'watch', 'start'].map(name => {
  jest.doMock(`../scripts/${name}`, () =>
    jest.fn(options => {
      currentScriptFile = name
      expect(options).toMatchSnapshot()
    })
  )
})
;['publish', 'release'].map(name => {
  jest.doMock(`../scripts/${name}`, () =>
    jest.fn(options => {
      const { cliArgs } = options
      currentScriptFile = name
      expect(`cozy-${name} ${cliArgs.join(' ')}`).toMatchSnapshot()
    })
  )
})

jest.doMock('../scripts/config', () =>
  jest.fn(options => {
    expect(options).toMatchSnapshot()
    return {
      config: 'Mocked config'
    }
  })
)

jest.doMock('../scripts/lint', () =>
  jest.fn(options => {
    const { cliArgs } = options
    currentScriptFile = 'lint'
    expect(`eslint ${cliArgs.join(' ')}`).toMatchSnapshot()
  })
)

jest.doMock('jest', () => ({
  run: jest.fn(args => {
    currentScriptFile = 'test'
    expect(`jest ${args.join(' ')}`).toMatchSnapshot()
  })
}))

// we reproduce the process.argv passed by node
// see https://nodejs.org/docs/latest/api/process.html#process_process_argv
function addCLIArgs(...args) {
  process.argv = process.argv.concat(args)
}

function callCLI() {
  jest.resetModules() // force doing require each time
  require('../bin/cozy-scripts')
}

describe('cozy-scripts (cs) CLI', () => {
  beforeEach(() => {
    jest.resetAllMocks()
    console.log = actualLog
    console.error = actualError
    currentScriptFile = null
    // clean args list
    process.argv = [process.execPath, path.resolve('../bin/cozy-scripts')]
    // default env
    delete process.env[CTS.HOT]
    delete process.env[CTS.DEBUG]
    delete process.env[CTS.ANALYZER]
    delete process.env[CTS.ESLINT_FIX]
    delete process.env[CTS.SRC_DIR]
    delete process.env[CTS.BUILD_DIR]
    delete process.env[CTS.MANIFEST]
  })

  afterAll(() => {
    // reset the argv environment variable
    process.argv = actualArgv.slice(0)
  })

  it('should display an help message if no args passed', () => {
    console.error = jest.fn()
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(stripAnsi(console.error.mock.calls[0][0])).toMatchSnapshot()
  })

  it('should display an error message if the action/command is not recognize', () => {
    console.error = jest.fn()
    // add cli arguments
    addCLIArgs('nobodyKnowIt', '--watch')
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(stripAnsi(console.error.mock.calls[0][0])).toMatchSnapshot()
  })

  it('should display the webpack config if option `--show-config` passed', () => {
    console.log = jest.fn()
    addCLIArgs('build', '--browser', '--show-config')
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(stripAnsi(console.log.mock.calls[0][0])).toMatchSnapshot()
  })

  it('should call jest on `test` command and pass remaining args to jest', () => {
    // add cli arguments
    addCLIArgs('test', '--watch', '--runInBand', '--config', 'undefile.js')
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(currentScriptFile).toBe('test')
  })

  it('should call eslint on `lint` command and pass remaining args to eslint', () => {
    // add cli arguments
    addCLIArgs('lint', '--fix', '{src,test}/**/*.{js,jsx}')
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(currentScriptFile).toBe('lint')
  })

  it('should call cozy-release on `release` command and pass remaining args to cozy-release', () => {
    // add cli arguments
    addCLIArgs('release', 'patch', '1.2.4')
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(currentScriptFile).toBe('release')
  })

  it('should call cozy-publish on `publish` command and pass remaining args to cozy-publish', () => {
    // add cli arguments
    addCLIArgs('publish', '--token', 'aMockedReGIStrYToKeN')
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(currentScriptFile).toBe('publish')
  })

  it('should build on `build` command with all passed options with production mode by default', () => {
    // add cli arguments
    addCLIArgs('build', '--browser')
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(currentScriptFile).toBe('build')
  })

  it('should watch on `watch` command with all passed options with development mode by default', () => {
    // add cli arguments
    addCLIArgs('watch', '--browser')
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(currentScriptFile).toBe('watch')
  })

  it('should handle --analyzer option', () => {
    // add cli arguments
    addCLIArgs('watch', '--browser', '--analyzer')
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(process.env[CTS.ANALYZER]).toBe('true')
  })

  it('should handle debug mode with --debug option', () => {
    // add cli arguments
    addCLIArgs('watch', '--debug')
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(process.env[CTS.DEBUG]).toBe('true')
  })

  it('should handle eslint fix mode with --fix option', () => {
    // add cli arguments
    addCLIArgs('lint', '--fix')
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(process.env[CTS.ESLINT_FIX]).toBe('true')
  })

  it('should handle hot reload mode with --hot option', () => {
    // add cli arguments
    addCLIArgs('watch', '--hot')
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(process.env[CTS.HOT]).toBe('true')
  })

  it('should handle changing dev server port with --port option', () => {
    // add cli arguments
    addCLIArgs('watch', '--port', '9999')
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(process.env[CTS.PORT]).toBe('9999')
  })

  it('should handle changing dev server host with --host option', () => {
    // add cli arguments
    addCLIArgs('watch', '--host', 'cozy.tools')
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(process.env[CTS.HOST]).toBe('cozy.tools')
  })

  it('should handle src directory path providing with --src-dir option', () => {
    // add cli arguments
    const customPath = 'custom/app/src'
    addCLIArgs('watch', '--src-dir', customPath)
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(process.env[CTS.SRC_DIR]).toBe(customPath)
  })

  it('should handle build directory path providing with --build-dir option', () => {
    // add cli arguments
    const customPath = 'custom/app/buildHere'
    addCLIArgs('watch', '--build-dir', customPath)
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(process.env[CTS.BUILD_DIR]).toBe(customPath)
  })

  it('should handle manifest path providing with --manifest option', () => {
    // add cli arguments
    const customPath = 'custom/app/manifest.webapp'
    addCLIArgs('watch', '--manifest', customPath)
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(process.env[CTS.MANIFEST]).toBe(customPath)
  })

  it('should handle config path providing with --config option', () => {
    // add cli arguments
    const customPath = 'custom/app/app.config.js'
    addCLIArgs('watch', '--config', customPath)
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(process.env[CTS.CONFIG]).toBe(customPath)
  })

  it('should handle --production option', () => {
    // add cli arguments
    addCLIArgs('watch', '--browser', '--production')
    expect(() => {
      callCLI()
    }).not.toThrow()
  })

  it('should handle --development option', () => {
    // add cli arguments
    addCLIArgs('watch', '--browser', '--development')
    expect(() => {
      callCLI()
    }).not.toThrow()
  })

  it('should handle mobile target with --mobile option', () => {
    // add cli arguments
    addCLIArgs('watch', '--mobile')
    expect(() => {
      callCLI()
    }).not.toThrow()
  })

  it('should run start script on `start` command with all passed options with development mode by default', () => {
    // add cli arguments
    addCLIArgs('start', '--stack')
    expect(() => {
      callCLI()
    }).not.toThrow()
    expect(currentScriptFile).toBe('start')
  })
})
