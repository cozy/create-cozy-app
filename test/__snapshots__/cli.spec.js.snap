// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cozy-scripts (cs) CLI should build on \`build\` command with all passed options with production mode by default 1`] = `
Object {
  "cliArgs": Array [
    "--browser",
  ],
  "mode": "production",
  "target": "browser",
}
`;

exports[`cozy-scripts (cs) CLI should call cozy-publish on \`publish\` command and pass remaining args to cozy-publish 1`] = `"cozy-publish --token aMockedReGIStrYToKeN"`;

exports[`cozy-scripts (cs) CLI should call cozy-release on \`release\` command and pass remaining args to cozy-release 1`] = `"cozy-release patch 1.2.4"`;

exports[`cozy-scripts (cs) CLI should call eslint on \`lint\` command and pass remaining args to eslint 1`] = `"eslint --fix {src,test}/**/*.{js,jsx}"`;

exports[`cozy-scripts (cs) CLI should call jest on \`test\` command and pass remaining args to jest 1`] = `"jest --watch --runInBand --config undefile.js"`;

exports[`cozy-scripts (cs) CLI should display an error message if the action/command is not recognize 1`] = `"cozy-scripts: unknown command nobodyKnowIt"`;

exports[`cozy-scripts (cs) CLI should display an help message if no args passed 1`] = `"Use \`--help\` option to get more informations about cozy-scripts usage."`;

exports[`cozy-scripts (cs) CLI should display the webpack config if option \`--show-config\` passed 1`] = `
Object {
  "cliArgs": Array [
    "--browser",
    "--show-config",
  ],
  "mode": "production",
  "target": "browser",
}
`;

exports[`cozy-scripts (cs) CLI should display the webpack config if option \`--show-config\` passed 2`] = `
"{
  \\"config\\": \\"Mocked config\\"
}"
`;

exports[`cozy-scripts (cs) CLI should handle --analyzer option 1`] = `
Object {
  "cliArgs": Array [
    "--browser",
    "--analyzer",
  ],
  "mode": "development",
  "target": "browser",
}
`;

exports[`cozy-scripts (cs) CLI should handle --development option 1`] = `
Object {
  "cliArgs": Array [
    "--browser",
    "--development",
  ],
  "mode": "development",
  "target": "browser",
}
`;

exports[`cozy-scripts (cs) CLI should handle --production option 1`] = `
Object {
  "cliArgs": Array [
    "--browser",
    "--production",
  ],
  "mode": "production",
  "target": "browser",
}
`;

exports[`cozy-scripts (cs) CLI should handle build directory path providing with --build-dir option 1`] = `
Object {
  "cliArgs": Array [
    "--build-dir",
    "custom/app/buildHere",
  ],
  "mode": "development",
  "target": "browser",
}
`;

exports[`cozy-scripts (cs) CLI should handle changing dev server host with --host option 1`] = `
Object {
  "cliArgs": Array [
    "--host",
    "cozy.tools",
  ],
  "mode": "development",
  "target": "browser",
}
`;

exports[`cozy-scripts (cs) CLI should handle changing dev server port with --port option 1`] = `
Object {
  "cliArgs": Array [
    "--port",
    "9999",
  ],
  "mode": "development",
  "target": "browser",
}
`;

exports[`cozy-scripts (cs) CLI should handle config path providing with --config option 1`] = `
Object {
  "cliArgs": Array [
    "--config",
    "custom/app/app.config.js",
  ],
  "mode": "development",
  "target": "browser",
}
`;

exports[`cozy-scripts (cs) CLI should handle debug mode with --debug option 1`] = `
Object {
  "cliArgs": Array [
    "--debug",
  ],
  "mode": "development",
  "target": "browser",
}
`;

exports[`cozy-scripts (cs) CLI should handle eslint fix mode with --fix option 1`] = `"eslint --fix"`;

exports[`cozy-scripts (cs) CLI should handle hot reload mode with --hot option 1`] = `
Object {
  "cliArgs": Array [
    "--hot",
  ],
  "mode": "development",
  "target": "browser",
}
`;

exports[`cozy-scripts (cs) CLI should handle manifest path providing with --manifest option 1`] = `
Object {
  "cliArgs": Array [
    "--manifest",
    "custom/app/manifest.webapp",
  ],
  "mode": "development",
  "target": "browser",
}
`;

exports[`cozy-scripts (cs) CLI should handle mobile target with --mobile option 1`] = `
Object {
  "cliArgs": Array [
    "--mobile",
  ],
  "mode": "development",
  "target": "mobile",
}
`;

exports[`cozy-scripts (cs) CLI should handle src directory path providing with --src-dir option 1`] = `
Object {
  "cliArgs": Array [
    "--src-dir",
    "custom/app/src",
  ],
  "mode": "development",
  "target": "browser",
}
`;

exports[`cozy-scripts (cs) CLI should run start script on \`start\` command with all passed options with development mode by default 1`] = `
Object {
  "cliArgs": Array [
    "--stack",
  ],
  "mode": "development",
  "stack": true,
  "target": "browser",
}
`;

exports[`cozy-scripts (cs) CLI should watch on \`watch\` command with all passed options with development mode by default 1`] = `
Object {
  "cliArgs": Array [
    "--browser",
  ],
  "mode": "development",
  "target": "browser",
}
`;
